HARK/core.py

```diff
@@ -862,207 +1021,303 @@ def clear_history(self):
         if not infinite_horizon:
+            # If finite horizon model, add cycle to the growing list
             solution = solution_cycle + solution
-
-        # Check for termination: identical solutions across
-        # cycle iterations or run out of cycles
-        solution_now = solution_cycle[0]
-        if infinite_horizon:
-            if completed_cycles > 0:
-                solution_distance = solution_now.distance(solution_last)
-                agent.solution_distance = (
-                    solution_distance  # Add these attributes so users can
-                )
-                agent.completed_cycles = (
-                    completed_cycles  # query them to see if solution is ready
-                )
-                go = (
-                    solution_distance > agent.tolerance
-                    and completed_cycles < max_cycles
-                )
-            else:  # Assume solution does not converge after only one cycle
-                solution_distance = 100.0
-                go = True
-        else:
             cycles_left += -1
             go = cycles_left > 0
-
-        # Update the "last period solution"
-        solution_last = solution_now
+            # Don't count replacement of terminal_partial as a cycle; see below
+            if solution_next.Bilt.stge_kind['iter_status'] == 'terminal_partial':
+                cycles_left += 1
+                completed_cycles += -1
+                go = True
+        else:  # infinite horizon
+            solution = solution_cycle
+            solution_now = solution_cycle[0]  # element 0 most recently solved
+            solution_now.solution_distance = \
+                solution_distance = solution_now.distance(solution_next)
+            go = (
+                solution_distance > agent.tolerance
+                and completed_cycles < max_cycles
+            )
+            # To permit ".solve()" method on the agent to resume solution after
+            # reaching convergence under some initial tolerance, user must
+            # * Change something about the problem (e.g., tolerance)
+            # * Change solution[0] stge_kind['iter_status'] to 'iterator'
+            # * Restart the solution process with [instance].solve()
+            # TODO: this has been tested only for PerfForesightConsumerType
+            # and IndShockConsumerType.  We should test whether agent is one of
+            # those before allowing resume
+
+            if hasattr(agent, 'solve_resume') and (agent.solve_resume is True):  # if resumption requested,
+                go = True  # solve one period/stage for sure, then keep going
+                agent.solve_resume = False  # go until stop criteria satisfied
+            if not go:  # Finished solving
+                # All models should incorporate 'stge_kind', but some have not
+                # Handle cases where that has not yet been implemented:
+                if not hasattr(solution_now.Bilt, 'stge_kind'):
```

Comment:
what is "Bilt"?

Reply:
It is a deliberate misspelling of "Built" - results that are constructed or might be modified in the course of the solution (as distinct from "Parm" which is parameters fed into the code).

PS.  I know your preference is for words that are deliberately spelled correctly.  A virtue of my choice of variable names is that if we can agree on a better alternative, a global search and replace will break neither normal text nor code.